File: adapter/src/main/scala/edu/leidenuniv/AuToBIAdapter/FeatureNamesBuilder.scala
package edu.leidenuniv.AuToBIAdapter

import edu.cuny.qc.speech.AuToBI.core.FeatureSet
import edu.leidenuniv.AuToBIAdapter.core.{FeatureExtractors, FeatureSets}
import scala.jdk.CollectionConverters.*
import upickle.default.{read, write}

class FeatureNamesBuilder(val featureSet: FeatureSet):
  def this() = this(FeatureSet())

  def withDefaultFeatures(name: String): this.type =
    FeatureSets.getFeatureSet(name) match
      case None    =>
        throw new IllegalAccessException(s"Cannot add features $name, not found")
      case Some(x) => x.getRequiredFeatures.forEach(featureSet.insertRequiredFeature(_))
    this

  def withFeature(name: String): this.type =
    if FeatureExtractors.checkFeaturePossible(name) then
      println(s"Adding: $name")
      featureSet.insertRequiredFeature(name)
    else
      println(s"cannot add $name")
      throw new IllegalAccessException(
        s"Cannot add feature $name, no extractor available",
      )
    this

  def withFeatures(names: List[String]): this.type =
    names.foreach(withFeature)
    this

  def withFeatures(names: String): this.type =
    withFeatures(read[List[String]](names))

  def build(): String =
    write(featureSet.getRequiredFeatures.asScala)

File: python/src/autobi/featurenames_builder.py
from __future__ import annotations

from typing import List

from py4j.java_gateway import Py4JJavaError
from autobi.core import AutobiJVM


class FeaturenamesBuilder:
    def __init__(self, jvm: AutobiJVM):
        self._jvm = jvm
        self._object = jvm._jvm.edu.leidenuniv.AuToBIAdapter.FeatureNamesBuilder()

    def with_default_features(self, name: str) -> FeaturenamesBuilder:
        try:
            self._object.withDefaultFeatures(name)
        except Py4JJavaError:
            raise ValueError(f"Cannot add feature {name}, not found")
        return self

    def with_feature(self, name: str) -> FeaturenamesBuilder:
        try:
            self._object.withFeature(name)
        except Py4JJavaError:
            raise ValueError(f"Cannot add feature {name}, not found")
        return self

    def with_features(self, names: List[str]) -> FeaturenamesBuilder:
        _name = "\t".join(names)
        try:
            self._object.withFeatures(_name)
        except Py4JJavaError:
            raise ValueError(f"Cannot add features {names}, not found")
        return self

    def build(self) -> List[str]:
        return str(self._object.build()).split("\t")
